pkg/controller/webhookbootstrap/controller.go

```diff
@@ -386,58 +395,60 @@ func (c *controller) updateSecret(secret *corev1.Secret, pk, ca, crt []byte) err
 	secret.Data[corev1.TLSPrivateKeyKey] = pk
 	secret.Data[corev1.TLSCertKey] = crt
 	secret.Data[cmmeta.TLSCAKey] = ca
-	_, err := c.kubeClient.CoreV1().Secrets(secret.Namespace).Update(secret)
+	secret.Type = corev1.SecretTypeTLS
+
+	// If the secret does not yet exist then we should create it
+	_, err := c.secretLister.Secrets(secret.Namespace).Get(secret.Name)
+	if apierrors.IsNotFound(err) {
+		_, err = c.kubeClient.CoreV1().Secrets(secret.Namespace).Create(secret)
+		return err
+	}
+
+	if err != nil {
+		return err
+	}
+
+	_, err = c.kubeClient.CoreV1().Secrets(secret.Namespace).Update(secret)
 	return err
 }
 
-// ensureSecretsExist ensures that the webhook secrets actually exist.
-// This is to ensure that the ProcessItem function is actually called with the
-// webhook's Secret resource, so that it can be provisioned.
-func (c *controller) ensureSecretsExist(ctx context.Context) {
+// bootstrapSecrets will manually add the Webhook's CA and Serving Secrets onto
+// the process queue. This is done since the Secrets have not been created yet.
+// We could create empty Secrets here instead to trigger reconciles however,
+// this causes issues whereby empty Secrets can delays in the webhook pod's
+// start. See #2537
+func (c *controller) bootstrapSecrets(ctx context.Context) {
```

Comment:
With the new implementation, we are now not only checking for the secrets existence every 10s, but also reading, decoding and parsing the 2 keypairs to ensure they are valid.

This doesn't strictly need to happen every 10s - if the Secret exists, the handler will be receiving resync requests anyway. Meaning that if the named Secret resource already exists, this 'bootstrap' method shouldn't need to do anything at all (it's only in the case where the Secret resource *doesn't* exist that we need to manually trigger a sync).

I am not sure how significant it'd be to change this up to accommodate that, nor do I have an idea of how 'expensive' this loop will be to run in practice, but thought it worth calling out 😄

Reply:
Do you want me to increase the time by 30s maybe?